Objects/exceptions.c

```diff
@@ -629,6 +629,454 @@ ComplexExtendsException(PyExc_BaseException, SystemExit, SystemExit,
                         0, 0, SystemExit_members, 0, 0,
                         "Request to exit from the interpreter.");
 
+/*
+ *    BaseExceptionGroup extends BaseException
+ *    ExceptionGroup extends BaseExceptionGroup and Exception
+ */
+
+PyObject *PyExc_ExceptionGroup;
+
+static inline PyBaseExceptionGroupObject*
+_PyBaseExceptionGroupObject_cast(PyObject *exc)
+{
+    assert(PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_BaseExceptionGroup));
+    return (PyBaseExceptionGroupObject *)exc;
+}
+
+static PyObject *
+BaseExceptionGroup_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    PyObject *msg = NULL;
+    PyObject *excs = NULL;
+
+    if (!PyArg_ParseTuple(args, "OO", &msg, &excs)) {
+        return NULL;
+    }
+    if (!PyUnicode_CheckExact(msg)) {
+        PyErr_SetString(PyExc_TypeError, "Expected a message");
+        return NULL;
+    }
+    Py_ssize_t numexcs = -1;
+    if (PySequence_Check(excs)) {
+        numexcs = PySequence_Length(excs);
+    }
+    if (numexcs <= 0) {
+        PyErr_SetString(PyExc_TypeError, "Expected a non-empty sequence");
+        return NULL;
+    }
+
+    int nested_base_exceptions = 0;
+    for (Py_ssize_t i = 0; i < numexcs; i++) {
+        PyObject *exc = PySequence_GetItem(excs, i);
+        if (!exc) {
+            return NULL;
+        }
+        if (!PyExceptionInstance_Check(exc)) {
+            PyErr_SetString(PyExc_TypeError,
+                "Nested exceptions must derive from BaseException");
```

Comment:
Can we mention at which index the non-exception was found? Also, "Nested exceptions" is not necessarily clear without context. Maybe "item N of second argument"?